lib/remap-lang/src/lib.rs

```diff
@@ -0,0 +1,140 @@
+mod error;
+mod expression;
+mod function;
+mod operator;
+mod parser;
+mod program;
+mod runtime;
+mod state;
+mod value;
+
+use expression::{Expr, Expression};
+use function::{Argument, ArgumentList, Function, Parameter};
+use operator::Operator;
+use state::State;
+
+pub use error::Error;
+pub use program::Program;
+pub use runtime::Runtime;
+pub use value::Value;
+
+type Result<T> = std::result::Result<T, Error>;
+
+/// Any object you want to map through the remap language has to implement this
+/// trait.
+pub trait Object: std::fmt::Debug {
+    /// Insert a given [`Value`] in the provided [`Object`].
+    ///
+    /// The `path` parameter determines _where_ in the given object the value
+    /// should be inserted.
+    ///
+    /// A path contains dot-delimited segments, and can contain a combination
+    /// of:
+    ///
+    /// * regular path segments:
+    ///
+    ///   ```txt
+    ///   .foo.bar.baz
+    ///   ```
+    ///
+    /// * quoted path segments:
+    ///
+    ///   ```txt
+    ///   .foo."bar.baz"
+    ///   ```
+    ///
+    /// * coalesced path segments:
+    ///
+    ///   ```txt
+    ///   .foo.(bar.baz | foobar | "bar.baz").qux
+    ///   ```
+    ///
+    /// * path indices:
+    ///
+    ///   ```txt
+    ///   .foo[2]
+    ///   ```
+    ///
+    /// When inserting into a coalesced path, the implementor is encouraged to
+    /// insert into the right-most segment if none exists, but can return an
+    /// error if needed.
+    fn insert(&mut self, path: &str, value: Value) -> std::result::Result<(), String>;
+
+    /// Find a value for a given path.
+    ///
+    /// See [`Object::insert`] for more details.
+    fn find(&self, path: &str) -> std::result::Result<Option<Value>, String>;
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
These are a few tests I used during development. In the next PR, this crate will be hooked up into Vector, and the existing tests there will use this crate, to prove nothing broke.

Once that's done, I'll move (most of) the unit tests over to this crate, and keep the behaviour tests in Vector.